### PR TITLE
ROX-10891: GitHub Actions for upstream release automation

### PR DESCRIPTION
**Pull request by 0x656b694d**
## Description

The initial PR on upstream automation. Covered some GitHub related activities, such as creating the release branch, milestones, cherry-picking and such.

Happy-path:
1. Start Release (manual action on GitHub);
2. Close RC milestone (the first milestone is created automatically);
3. Check CI runs (not automated);
4. Edit the published pre-release (RC) to be the final release with the suggested command.

The idea is to automate the steps which can be automated, suggest what to do next and notify about taken steps.
Workflow steps have to be restartable with no unexpected effects.

Steps summary is printed under the workflow run on GitHub for the release managers and the notifications for wider audience are sent to Slack.

*See ROX-10891 ticket for the link to the more detailed documentation*.

## Notes

The Slack channel ID points to my personal chat.
The long-running cluster creation is requested in the workflow, but no further action is taken.
Further work is required on the workflows to improve user communication and test cluster creation.

## Testing Performed

The Dry-run option implemented with which the jobs would only do local changes and print the progress without affecting the repos or GitHub objects.
Manual tests performed on https://github.com/stackrox/test-gh-actions repository.

### CODE REVIEW

File: .github/workflows/cut-rc.yml [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+      - name: Check open PRs
+        id: check
+        run: |
+          PRs=$(gh pr list -s open \
+          --repo "${{github.repository}}" \
+          --search "milestone:${{needs.variables.outputs.milestone}}" \
+          --json number \
+          --jq length)
+          if [ "$PRs" -gt 0 ]
+          then
```
Comment (vikin91) [inline .github/workflows/cut-rc.yml:105]: I usually prefer the condensed version of the if:

```suggestion
          if [ "$PRs" -gt 0 ]; then
```

File: .github/workflows/cut-rc.yml [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+            echo "Milestone ${{needs.variables.outputs.next-milestone}} has been created." >> $GITHUB_STEP_SUMMARY
+          fi
+
+      - name: Move open PRs
+        run: |
+          for PR in $(gh pr list -s open \
+            --repo "${{github.repository}}" \
+            --search "milestone:${{needs.variables.outputs.milestone}}" \
+            --json number \
+            --jq ".[] | .number")
```
Comment (vikin91) [inline .github/workflows/cut-rc.yml:134]: This may work fine, but it has some drawbacks.
I would suggest to capture the command output into an array and then iterate over that array in the for loop. This would allow us catch potential mistakes and handle properly edge-cases (e.g., empty values).

File: .github/workflows/cut-rc.yml [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+          do
+            [ "${{github.event.inputs.dry-run}}" != "true" ] && \
```
Comment (vikin91) [inline .github/workflows/cut-rc.yml:136]: Why don't we move this check to the top of this block - outside of the loop?

File: .github/workflows/cut-rc.yml [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+        id: cherry-pick
+        run: |
+          SLACK_MESSAGE_FILE=$(mktemp)
+          EMPTY=""
+          gh pr list -s closed \
+            --search "milestone:${{needs.variables.outputs.milestone}}" \
+            --base "${{env.main_branch}}" \
+            --json number,url,mergeCommit,author,title \
+            --jq '.[] | "\(.number)\t\(.url)\t\(.mergeCommit.oid)\t\(.author.login)\t\(.title)"' \
+          | while read -r PR_COMMIT
```
Comment (vikin91) [inline .github/workflows/cut-rc.yml:171]: Again separating the command from the loop would allow us to do some debugging in case the command returns something unexpected.

File: .github/workflows/cut-rc.yml [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+          git push
+          if [ -n "$EMPTY" ]; then
+            echo "Cherry-picked commits have been pushed to the release branch." >> $GITHUB_STEP_SUMMARY
+          fi
+
+          # Replace % with %25 and join lines with %0A:
+          FAILED="$(sed ':a; N; $!ba; s/%/%25/g; s/\n/%0A/g' $SLACK_MESSAGE_FILE)"
+          rm -f "$SLACK_MESSAGE_FILE"
+
+          if [ -n "$FAILED" ]
```
Comment (vikin91) [inline .github/workflows/cut-rc.yml:217]: I do not understand how the failed sed on a slack-message file is used to determine whether cherry-picking has succeeded. Could we think of providing some more robust check?

**Review by vikin91 — CHANGES_REQUESTED** [review]
This is a great piece of great work - thank you once again for starting this!

So far I have just went though the code (I still want to try it out), but I am sharing my feedback already now.

I think the code in the current shape needs some cleaning and reorganization. I suggest the following as the minimum:

1. Extract all bash code into functions and place then in a separate file that can be sourced.
2. Apply a dose of defensive programing (verify each input and output for correctness) with chatty logging, so that the debugging is easy
3. Source the bash file in each action and use the functions - this will keep the yaml size minimal
4. If possible and reasonable, cover the library with bats tests.

As this may be a relatively large effort, I am happy to help you with that :)

File: .github/workflows/cut-rc.yml (hunk 5 above)
Comment (0x656b694d) [reply .github/workflows/cut-rc.yml:217]: SLACK_MESSAGE_FILE file is basically a list of failed PRs to be reported to SLACK. /* I could probably use bash arrays, but I avoid advanced bash features as you could have noticed */
So if it is empty, there were no merge failures. I considered it to be a good check.

**Comment by 0x656b694d** [comment]
> 1. Extract all bash code into functions and place then in a separate file that can be sourced.

This would require `git clone` of the repository, which is often an overkill for a couple of commands. I'd suggest to move only complex code to scripts (like deploying long-running cluster) and keep fewliners inlined.

> 2. Apply a dose of defensive programing (verify each input and output for correctness) with chatty logging, so that the debugging is easy

Agree.

> 3. Source the bash file in each action and use the functions - this will keep the yaml size minimal

Don't fully agree.

> 4. If possible and reasonable, cover the library with bats tests.

All the actions have side effects, not sure we want to mock all the external services to test how we call their APIs. But I'm open for reasonable testing suggestions.

File: .github/workflows/create-cluster.yml [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      - name: Create Cluster
+        env:
+          INFRA_TOKEN: ${{secrets.INFRA_TOKEN}}
+        run: |
+          NAME="${{inputs.name}}"
+          WAIT=""
+          if [ "${{inputs.wait}}" = "true" ]
+          then
+            WAIT="--wait";
+            echo "::warning::The job will wait for the cluster creation."
```
[the remark this answers is not in the document]
Comment (porridge) [reply .github/workflows/create-cluster.yml:49]: I'd also suggest leaving at least the ` to finish` part for clarity.

File: .github/workflows/cut-rc.yml [hunk 7]
```diff
@@ -0,0 +1,547 @@
...
+        description: Dry-run
+        required: false
+        default: true
+        type: boolean
+
+env:
+  main_branch: master
+  docs_repository: openshift/openshift-docs
+  jira_projects: ROX, RS, RTOOLS
+  slack_channel: C03D78P4RFF
```
Comment (porridge) [inline .github/workflows/cut-rc.yml:24]: How about a comment which channel this is?
Also, should we change this to something visible to others now?

File: .github/workflows/cut-rc.yml [hunk 8]
```diff
@@ -0,0 +1,547 @@
...
+
+              # Create PR to the release branch:
+              gh create pr --base "${{needs.variables.outputs.branch}}" --fill \\
+                --milestone "${{needs.variables.outputs.milestone}}"
+
+          EOF
+            exit 1
+          fi
+      - name: Post to Slack
+        if: always() && steps.cherry-pick.outputs.bad-cherries != ''
```
Comment (porridge) [inline .github/workflows/cut-rc.yml:262]: `if: always()`?
What does it mean 🤯 ?

File: .github/workflows/cut-rc.yml [hunk 9]
```diff
@@ -0,0 +1,547 @@
...
+      - name: Initialize mandatory git config
+        run: |
+          git config user.name "${{github.event.sender.login}}"
+          git config user.email noreply@github.com
+      - name: Update docs submodule
+        run: |
+          git submodule update --init --remote -- docs/content
+          git add docs/content
+          if ! git diff-index --quiet HEAD; then
+            git commit -m "Update docs submodule"
```
Comment (porridge) [inline .github/workflows/cut-rc.yml:294]:
```suggestion
            git commit -m "Update docs submodule for milestone ${{needs.variables.outputs.milestone}}"
```
? (no idea whether that would actually work)

File: .github/workflows/create-cluster.yml [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+      lifespan:
+        description: Cluster lifespan
+        type: string
+        default: 48h
+        required: false
+      args:
+        description: Comma separated flavor arguments. Ex. nodes=5,main-image=main:tag
+        type: string
+        required: false
+        # Spaces in arguments are not supported.
```
[the remark this answers is not in the document]
Comment (0x656b694d) [reply .github/workflows/create-cluster.yml]: Now supported. Comment removed.

File: .github/workflows/create-cluster.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (0x656b694d) [reply .github/workflows/create-cluster.yml:48]: Ok!

File: .github/workflows/create-cluster.yml [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        default: false
+jobs:
+  infra:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Download infractl
+        run: |
+          curl -sL https://infra.rox.systems/v1/cli/linux/amd64/upgrade \
+          | jq -r ".result.fileChunk" \
+          | base64 -d \
```
[the remark this answers is not in the document]
Comment (0x656b694d) [reply .github/workflows/create-cluster.yml:36]: Unfortunately in this case `jq` fails to decode the output.

File: .github/workflows/create-cluster.yml [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        type: boolean
+        required: false
+        default: false
+jobs:
+  infra:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Download infractl
+        run: |
+          curl -sL https://infra.rox.systems/v1/cli/linux/amd64/upgrade \
```
[the remark this answers is not in the document]
Comment (0x656b694d) [reply .github/workflows/create-cluster.yml:34]: `set -e` is implicitly set, I'll see if `ou` would help.
Will add the missing `--fail` here, thanks.

File: .github/workflows/cut-rc.yml (hunk 7 above)
Comment (0x656b694d) [reply .github/workflows/cut-rc.yml:24]: This is now the `epic-release-improvements` channel. We'll change it to `eng-release` eventually. Well, I could add that in a comment.

File: .github/workflows/cut-rc.yml (hunk 8 above)
Comment (0x656b694d) [reply .github/workflows/cut-rc.yml:267]: github actions magic, this is required to not skip the step after previous step failure. Could probably be changed to `failure()`.

**Comment by 0x656b694d** [comment]
@porridge, thanks for the review!
With GitHub actions we've got the following possibilities for code reuse:
* Call external action (ex. in the PR: checkout code, slack, install kubectl);
* Call another workflow (ex. in the PR: create-clusters.yml, variables.yml);
* Call local script (ex. in the PR: ./deploy.sh)

Please highlight the code which could have been extracted for reuse so we could move it to an appropriate module.

**Comment by porridge** [comment]
> Please highlight the code which could have been extracted for reuse so we could move it to an appropriate module.

I think anything longer than, say, 5 lines is a candidate for extraction 🤔

**Comment by 0x656b694d** [comment]
> I think anything longer than, say, 5 lines is a candidate for extraction 

When extracting the code, we'll need to consider the (time) cost of repository cloning for calling a script from the repo, and/or instantiating another job runner. Working on a branch we'll probably want to call scripts from the main branch (as from where the workflow is running), which may also add delays and complexity.
That may add seconds to the procedure that a person would monitor, which may not always be beneficial for user experience.

File: .github/workflows/cut-rc.yml [hunk 13]
```diff
@@ -0,0 +1,550 @@
...
+
+  create-k8s-cluster:
+    name: Create k8s cluster
+    needs: [variables, cut-rc]
+    if: github.event.inputs.dry-run != 'true'
+    uses: ./github/workflows/create-cluster.yml
+    with:
+      flavor: qa-demo
+      name: qa-k8s-${{needs.variables.outputs.milestone}}
+      args: main-image=main:${{needs.variables.outputs.milestone}}
```
Comment (0x656b694d) [inline .github/workflows/cut-rc.yml:375]:
```suggestion
      args: main-image=docker.io/stackrox/main:${{needs.variables.outputs.milestone}}
```

**Review by 0x656b694d — COMMENTED** [review]
Set the full main image name.

**Comment by vikin91** [comment]
Sorry, I didn't make to give it another look :( I will do it as soon as I feel a bit better

**Review by sbostick — COMMENTED** [review]
### Overall impression and cursory analysis

* Nicely captures the steps and dependencies
* Clear separation of inputs, outputs, secrets
* Readable and maintainable
* Deps (cli): curl, gh, infractl, git
* Deps (gha): actions/checkout@v3, slackapi/slack-github-action@v1.19.0, azure/setup-kubectl@v2.0
* Runtime: ubuntu-latest, bash

**Comment by sbostick** [comment]
### Suggestions

Add a README.md to the effect:

Implements ACS upstream release branching model including:
* targeted issue state verification
* pr creation
* test env provisioning

References:
* [Release Mangement - upstream](https://docs.engineering.redhat.com/display/StackRox/How+to+manage+a+release)
* [Release Management - downstream](https://docs.engineering.redhat.com/display/StackRox/Managing+a+Release+on+Red+Hat+Infrastructure)
* [Release Process](https://docs.engineering.redhat.com/display/StackRox/Managing+a+Release+on+Red+Hat+Infrastructure)
* [Release Tracker](https://docs.google.com/spreadsheets/d/19HaplE_vNOUFN2wrkQRf-hNXNspajaOA2mgXMdjyRPs/edit#gid=769572822)

**Comment by sbostick** [comment]
### Possible improvements

I agree with Piotr on script reuse if possible, and generally, factoring out shell code from yaml. However in this case I found it quite readable and self-documenting as is, so no strong preference from me. I'd say the current implementation is a good evolution of the long-form instructions.

So factoring out into scripts could improve portability and testability, but I'm not sure I could justify the ROI. And based on what you captured into code already, if one was motivated to reimplement, it would be a fairly straightfoward translation. For example, I'm most comfortable in Python and would prefer being able to mock out external API calls and add tests for the actual logic and corner cases related to branching/pr/milestone/issue manipulation, empty search results, etc. Another option might be making an e2e test of the release model, essentially treating it as a strategy that can be reused. But designing and testing a more generalized, reusable solution is sort of outside the scope. At least could defer larger refactoring and/or generalization efforts to separate initiatives with their own clear motivations and goals.

File: .github/workflows/cut-rc.yml [hunk 14]
```diff
@@ -0,0 +1,598 @@
...
+        description: Dry-run
+        required: false
+        default: true
+        type: boolean
+
+env:
+  main_branch: master
+  docs_repository: openshift/openshift-docs
+  jira_projects: ROX, RS, RTOOLS
+  slack_channel: C03KSV3N6N8 #test-release-automation
```
Comment (JoukoVirtanen) [inline .github/workflows/cut-rc.yml:24]: I think we already have this in CircleCI config in plain text. Is it possible to put this in a secret?

Comment (0x656b694d) [reply .github/workflows/cut-rc.yml:24]: It is possible. Benefit: single source for all workflows. Disadvantage: only admins have access. As the ID is not necessarily a sensitive information, we might want to keep during testing phase.

File: .github/workflows/cut-rc.yml [hunk 15]
```diff
@@ -0,0 +1,598 @@
...
+
+          ./deploy/k8s/sensor.sh
+
+          kubectl -n stackrox set env deploy/sensor MUTEX_WATCHDOG_TIMEOUT_SECS=0
+          kubectl -n stackrox set env deploy/sensor ROX_FAKE_KUBERNETES_WORKLOAD=long-running
+          kubectl -n stackrox patch deploy/sensor -p '{"spec":{"template":{"spec":{"containers":[{"name":"sensor","resources":{"requests":{"memory":"3Gi","cpu":"2"},"limits":{"memory":"12Gi","cpu":"4"}}}]}}}}'
+
+          kubectl -n stackrox set env deploy/central MUTEX_WATCHDOG_TIMEOUT_SECS=0
+          kubectl -n stackrox patch deploy/central -p '{"spec":{"template":{"spec":{"containers":[{"name":"central","resources":{"requests":{"memory":"3Gi","cpu":"2"},"limits":{"memory":"12Gi","cpu":"4"}}}]}}}}'
+          echo "Long-running GKE cluster has been patched" >> $GITHUB_STEP_SUMMARY
```
Comment (JoukoVirtanen) [inline .github/workflows/cut-rc.yml:556]: When I created my script to create the long-running cluster I had to make some changes to have it wait for things. When copying and pasting these commands from the documentation it works, because there is time for things to finish, but if you put everything in a single scripts you need to put checks to make sure that the previous step completed. I think you will eventually replace this with my script, so it is fine.

**Comment by JoukoVirtanen** [comment]
This is great work! Thanks for doing this! I thought that automating everything that you have automated here would take a long time and be done in many PRs, but you are doing it in one. My main criticism is that you have shell in the config files for things that we already have script files for. This is an issue that I think can be addressed in future PRs.

File: .github/workflows/cut-rc.yml [hunk 16]
```diff
@@ -0,0 +1,598 @@
...
+      JIRA_TOKEN: ${{ secrets.JIRA_TOKEN }}
+    steps:
+      - name: Query JIRA
+        env:
+          JQL: >-
+            project IN (${{env.jira_projects}})
+            AND fixVersion = \"${{needs.variables.outputs.release-patch}}\"
+            AND status != CLOSED
+            AND Component != Documentation AND type != Epic
+            order by assignee
```
[the remark this answers is not in the document]
Comment (0x656b694d) [reply .github/workflows/cut-rc.yml:47]: Using script would require code checkout here, which is acceptable as there will be no confusion with the release branch. This code also post comments in the found Jira tickets to grab attention of the assignees.

**Review by rukletsov — APPROVED** [review]
I think the suggestion to pull some part of the workflows into scripts is a reasonable one. There is an interleaved dependency on CircleCI, ideally, we should simplify them to the following: run desired CI checks and trigger a release after. I believe this (i.e. simplification) will affect the workflows here.

So let's maybe start with what we have, polish it during the upcoming release, and factor out _after_ we shuffle steps between CI and release for simplicity.

**Review by sbostick — APPROVED** [review]

File: .github/workflows/cut-rc.yml (hunk 3 above)
Comment (vikin91) [reply .github/workflows/cut-rc.yml:144]: I do not understand your answer. If the `"${{github.event.inputs.dry-run}}"` is set to `true`, then nothing will be run from the lines 144-150. What is the point in checking `[ "${{github.event.inputs.dry-run}}" != "true" ]` multiple times when once would be enough?

File: .github/workflows/cut-rc.yml [hunk 17]
```diff
@@ -0,0 +1,503 @@
...
+            --jq ".[] | .number")
+          do
+            [ "${{github.event.inputs.dry-run}}" != "true" ] && \
+            gh pr edit $PR \
+              --milestone "${{needs.variables.outputs.next-milestone}}" \
+              --repo "${{github.repository}}"
+            echo "PR $PR has been moved to milestone ${{needs.variables.outputs.next-milestone}}." >> $GITHUB_STEP_SUMMARY
+          done
+
+  check-merged-prs:
```
[the remark this answers is not in the document]
Comment (vikin91) [reply .github/workflows/cut-rc.yml:152]: I think, we should extract it on the next occasion. This should apply to all the bash code used in the yamls.

**Review by vikin91 — APPROVED** [review]
It looks good, so I am happy to accept it the way it is.

However, I find it important to make the bash code easier to manage. I think we should follow up on the idea of an external script library as one of the next things to work on.

File: .github/workflows/cut-rc.yml (hunk 3 above)
Comment (0x656b694d) [reply .github/workflows/cut-rc.yml:144]: I don't understand your question :). The current check disables only the `gh pr edit` command. The following `echo` need to be always printed in the loop. So I need this check inside the loop.

File: .github/workflows/create-cluster.yml [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+        type: boolean
+        required: false
+        default: false
+jobs:
+  infra:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Download infractl
+        run: |
+          curl --fail -sL https://infra.rox.systems/v1/cli/linux/amd64/upgrade \
```
Comment (janisz) [inline .github/workflows/create-cluster.yml:34]: How about adding `-S`?

>-S, --show-error
 >      When used with -s it makes curl show an error message if it fails.
